include/nbl/builtin/hlsl/bxdf/base/cook_torrance_base.hlsl

```diff
@@ -0,0 +1,378 @@
+        }
+        else
+            valid = _sample.getNdotL() > numeric_limits<scalar_type>::min && interaction.getNdotV() > numeric_limits<scalar_type>::min;
+        if (!valid)
+            return hlsl::promote<spectral_type>(0.0);
+
+        using quant_query_type = typename ndf_type::quant_query_type;
+        scalar_type dummy;
+        quant_query_type qq = ndf.template createQuantQuery<MicrofacetCache>(cache, dummy);
+
+        quant_type D = ndf.template D<sample_type, Interaction, MicrofacetCache>(qq, _sample, interaction, cache);
+        scalar_type DG = D.projectedLightMeasure;
+        if (D.microfacetMeasure < bit_cast<scalar_type>(numeric_limits<scalar_type>::infinity))
+        {
+            using g2g1_query_type = typename ndf_type::g2g1_query_type;
+            g2g1_query_type gq = ndf.template createG2G1Query<sample_type, Interaction>(_sample, interaction);
+            DG *= ndf.template correlated<sample_type, Interaction>(gq, _sample, interaction);
+        }
+        scalar_type clampedVdotH = cache.getVdotH();
+        NBL_IF_CONSTEXPR(IsBSDF)
+            clampedVdotH = hlsl::abs(clampedVdotH);
+        return impl::__implicit_promote<spectral_type, typename fresnel_type::vector_type>::__call(_f(clampedVdotH)) * DG;
+    }
+    template<typename C=bool_constant<!IsAnisotropic> >
+    enable_if_t<C::value && !IsAnisotropic, spectral_type> eval(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(isocache_type) cache)
+    {
+        return __eval<isotropic_interaction_type, isocache_type>(_sample, interaction, cache);
+    }
+    spectral_type eval(NBL_CONST_REF_ARG(sample_type) _sample, NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(anisocache_type) cache)
+    {
+        return __eval<anisotropic_interaction_type, anisocache_type>(_sample, interaction, cache);
+    }
+
+    template<typename C=bool_constant<!IsBSDF> >
+    enable_if_t<C::value && !IsBSDF, sample_type> generate(NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, const vector2_type u, NBL_REF_ARG(anisocache_type) cache)
+    {
+        if (interaction.getNdotV() > numeric_limits<scalar_type>::min)
+        {
+            ray_dir_info_type invalidL;
+            invalidL.makeInvalid();
+            return sample_type::createFromTangentSpace(invalidL, interaction.getFromTangentSpace());
+        }
+
+        const vector3_type localV = interaction.getTangentSpaceV();
+        const vector3_type localH = ndf.generateH(localV, u);
+        const scalar_type VdotH = hlsl::dot(localV, localH);
+        const vector3_type H = hlsl::mul(interaction.getFromTangentSpace(), localH);
+
+        ray_dir_info_type L;
+        if (scalar_type(2.0) * VdotH * localH.z > localV.z) // NdotL>0, compiler's Common Subexpression Elimination pass should re-use 2*VdotH later
+        {
+            assert(VdotH >= scalar_type(0.0));
```

Comment:
you can asset this as soon as `VdotH` is known with a comment about VNDF sampling